src/core/iomgr/fd_posix.c

```diff
@@ -184,7 +184,9 @@ void grpc_fd_orphan(grpc_fd *fd, grpc_iomgr_cb_func on_done, void *user_data) {
   fd->on_done_user_data = user_data;
   shutdown(fd->fd, SHUT_RDWR);
   ref_by(fd, 1); /* remove active status, but keep referenced */
+  gpr_mu_lock(&fd->watcher_mu);
   wake_all_watchers(fd);
+  gpr_mu_unlock(&fd->watcher_mu);
```

Comment:
Good call, renamed.

This is something that we should have testing infrastructure to catch - reviews can only go so far, and machines are much better at this than we are.

@nicolasnoble is on it though :)